### PR TITLE
Fix #626: Crash in CompilationWithAnalyzers.GetEffectiveDiagnostics

### LINKED ISSUE #626: Crash in CompilationWithAnalyzers.GetEffectiveDiagnostics
here are call stack.

[7376] System.InvalidOperationException: Collection was modified; enumeration operation may not execute. 
[7376]    at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource) 
[7376]    at System.Collections.Generic.List`1.Enumerator.MoveNextRare() 
[7376]    at System.Collections.Generic.List`1.Enumerator.MoveNext() 
[7376]    at System.Linq.Enumerable.WhereListIterator`1.MoveNext() 
[7376]    at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.<GetEffectiveDiagnostics>d__8.MoveNext() in D:\dd\FixMain\Open\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 79 
[7376]    at System.Linq.Buffer`1..ctor(IEnumerable`1 source) 
[7376]    at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source) 
[7376]    at System.Collections.Immutable.ImmutableArray.CreateRange[T](IEnumerable`1 items) 
[7376]    at System.Collections.Immutable.ImmutableArray.ToImmutableArray[TSource](IEnumerable`1 items) 
[7376]    at Microsoft.CodeAnalysis.Diagnostics.EngineV1.DiagnosticAnalyzerDriver.<GetSemanticDiagnosticsAsync>d__39.MoveNext() in D:\dd\FixMain\Open\src\Features\Core\Diagnostics\EngineV1\DiagnosticAnalyzerDriver.cs:line 396 
[7376] --- End of stack trace from previous location where exception was thrown --- 
[7376]    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) 
[7376]    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) 
[7376]    at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult() 
[7376]    at Microsoft.CodeAnalysis.Diagnostics.EngineV1.DiagnosticIncrementalAnalyzer.<GetSemanticDiagnosticsAsync>d__65.MoveNext() in D:\dd\FixMain\Open\src\Features\Core\Diagnostics\EngineV1\DiagnosticIncrementalAnalyzer.cs:line 849

### PR DESCRIPTION
**Pull request by mavasani**
Got: 'System.InvalidOperationException: Collection was modified; enumeration operation may not execute.'

Fix the GetEffectiveDiagnostics API to take an ImmutableArray<Diagnostic> instead of IEnumerable<Diagnostic>

@srivatsn @JohnHamby please take a look..

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs [hunk 1]
```diff
@@ -71,9 +71,9 @@ private void AddExceptionDiagnostic(Diagnostic diagnostic)
...
         /// 4) Pragma directives for the given <paramref name="compilation"/>.
         /// </summary>
-        public static IEnumerable<Diagnostic> GetEffectiveDiagnostics(IEnumerable<Diagnostic> diagnostics, Compilation compilation)
+        public static IEnumerable<Diagnostic> GetEffectiveDiagnostics(ImmutableArray<Diagnostic> diagnostics, Compilation compilation)
         {
-            if (diagnostics == null)
+            if (diagnostics == default(ImmutableArray<Diagnostic>))
             {
                 throw new ArgumentNullException(nameof(diagnostics));
             }
```
Comment (tmeschter) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs]: As `ImmutableArray` is a value type I don't think the `ArgumentNullException` is appropriate or necessary. Couldn't we simply treat `default(ImmutableArray<Diagnostic>)` as though we had an empty `ImmutableArray`?

File: src/Compilers/Core/Portable/PublicAPI.txt [hunk 2]
```diff
@@ -1899,7 +1899,7 @@ static Microsoft.CodeAnalysis.Diagnostic.Create(Microsoft.CodeAnalysis.Diagnosti
...
-static Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.GetEffectiveDiagnostics(System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.Diagnostic> diagnostics, Microsoft.CodeAnalysis.Compilation compilation)
+static Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.GetEffectiveDiagnostics(System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Diagnostic> diagnostics, Microsoft.CodeAnalysis.Compilation compilation)
 static Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.IsDiagnosticAnalyzerSuppressed(Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer analyzer, Microsoft.CodeAnalysis.CompilationOptions options, System.Func<System.Exception, Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer, bool> continueOnAnalyzerException)
```
Comment (tmeschter) [inline src/Compilers/Core/Portable/PublicAPI.txt]: Needs API review.

File: src/Features/Core/Diagnostics/EngineV1/DiagnosticAnalyzerDriver.cs [hunk 3]
```diff
@@ -309,7 +309,7 @@ private IEnumerable<Diagnostic> GetFilteredDocumentDiagnosticsCore(IEnumerable<D
 
             return compilation == null
                 ? diagnostics
-                : CompilationWithAnalyzers.GetEffectiveDiagnostics(diagsFilteredByLocation, compilation);
+                : CompilationWithAnalyzers.GetEffectiveDiagnostics(diagsFilteredByLocation.ToImmutableArray(), compilation);
```
Comment (tmeschter) [inline src/Features/Core/Diagnostics/EngineV1/DiagnosticAnalyzerDriver.cs]: Why `ToImmutableArray` here, and `ToImmutableArrayOrEmpty` elsewhere?

**Comment by srivatsn** [comment]
I don't understand this change - from the exception it sounds like some code was trying to mutate the collection. If we don't change that code and pass in an immutable array, that code will simply produce a new array - is that what we want? What was the offending piece of code?

File: src/Features/Core/Diagnostics/EngineV1/DiagnosticAnalyzerDriver.cs (hunk 3 above)
Comment (mavasani) [inline src/Features/Core/Diagnostics/EngineV1/DiagnosticAnalyzerDriver.cs]: Here we know that the value is non-null. I used to ImmutableArrayOrEmpty within the test extension method as I don't know if some caller passes in null.

File: src/Compilers/Core/Portable/PublicAPI.txt (hunk 2 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/PublicAPI.txt]: + @AnthonyDGreen 
Should I tag anyone else for API review? Do we have any recommended process around it?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs (hunk 1 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs]: Makes sense, I'll change this to just return the passed in array for this case. 
I think we need a similar change even for some of our Context registration methods which take an ImmutableArray as an argument and have a similar ArgumentNullException for default ImmutableArray. I'll update those as well.

Comment (tmeschter) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs]: You might want to dig into the differences between `default(ImmutableArray<...>)` and `ImmutableArray<...>.Empty`. I'm not sure if they are the same thing, or not.

**Comment by mavasani** [comment]
Sorry, I was too lazy to type out the whole context. This API is currently only used by the IDE driver when trying to filter our analyzer diagnostics per document/project. The DiagnosticAnalyzerDriver.Get*DiagnosticsAsync methods use a pooled list of diagnostics to register analyzer diagnostics, and pass that into the GetEffectiveDiagnostics method, returning the value returned by this method to its caller. I presume the pooled list object got collected before the source enumerable could be processed, causing this exception. I changed the DiagnosticAnalyzerDriver APIs to return an immutable array last week, so we shouldn't see this exception with current driver code.
However, given this is a public API, we shouldn't be iterating a passed in IEnumerable, as there is nothing that can prevent such mistakes by other API consumers. Probably the API should also return an ImmutableArray, but that is not a required change such as the parameter type.

**Comment by mavasani** [comment]
Btw, alternate fix here to avoid public API change is to just convert passed in IEnumerable to an ImmutableArray for iterating, not sure what is more preferable. I don't think we would have many consumers of this API currently though, do we? So even an API change shouldn't be that bad

**Comment by JohnHamby** [comment]
I would be very surprised if anyone outside of MS has taken a dependency on this API.

I'm not aware of any guidance against having IEnumerable parameters in APIs. It's more accepting than a specific collection type, which might counterbalance the possibility of having late failures like the once-possible one discussed above.

**Comment by mavasani** [comment]
@JohnHamby fair point. I think given that the root cause of the problem has been fixed in the IDE driver, changing or fixing this API is not urgent. John, would it be ok if I park this issue onto you, so that you can accumulate this possible API change OR fix as part of your driver refactoring prototype work I presume that is going to consumer this API? I'll also move this out of rc2.

**Comment by mavasani** [comment]
Discussed this with Sri. I am going to resolve this issue as fixed without merging this PR and file a separate bug for API fix/change issue.

**Comment by mavasani** [comment]
#802 tracks the API work now.

**Pull request closed without merging**